### ISSUE TITLE
How can we use plugins in the tailwind config?

### ISSUE THREAD

**Issue by wkd-kapsule** [remcohaszing/monaco-tailwindcss#92]
Hi!

First of all, I have to thank you for this project. At first I was sure that I'd never get Tailwind Intellisense in Monaco, given how difficult it is to configure anything in it. Then I discovered this repo and it felt like magic.

My only concern is the following: how can we configure the `worker` to handle custom classes added via `plugins`? In order for them to be picked up by autocompletion and hover providers. Something like this with "vanilla" `tailwind`:

``` typescript
const tailwindConfig = {
  // ... rest of the config
  plugins: [
    ({addUtilities}) => {
      addUtilities({
        '.text-accent': {color: 'var(--brand-12)'},
        '.text-muted': {color: 'var(--gray-11)'},      
      })
    }
  ]
}

configureMonacoTailwindcss(monaco, {tailwindConfig})
```

If I try to pass a `config` with such a `plugin`, I'll get an error from `postMessage` as it can't serialize/clone the `function` upon passing the `config` to the `worker`.

After digging into the source code of your library, I've found that you have an implementation for the different functions, as `addUtilities` or `addComponents`, behind an `api` object.

The ideal would be that the exposed `configureMonacoTailwindcss` could take more `options` for each `function` mentioned above.
It would give something like this:

This way, we would have access to these inner functions that we can't just directly pass to the `worker`.
``` typescript
const tailwindConfig = {
  // ...
  plugins: [],
}

const api = {
  addUtilities: {
    '.text-accent': {color: 'var(--brand-12)'},
    '.text-muted': {color: 'var(--gray-11)'},      
  }
  // addBase, addDefaults, addComponents, ...
}

configureMonacoTailwindcss(monaco, {tailwindConfig, api})
```

Hoping I can get some help, thanks again for your great work on this project.

NOTE: I've also tried to play with the `initialize` function that you expose, but it didn't do the trick. Quite frankly, I don't think it did anything at all. I struggle to understand the point of the function. The `prepareTailwindConfig` says it takes "the Tailwind configuration passed from the main thread". Is it the config passed to `configureMonacoTailwindcss` or do I miss something? And if so, what's the point of that function if we're already passing the `config` to `configureMonacoTailwindcss`?

**Comment by remcohaszing** [comment]
> First of all, I have to thank you for this project.

:heart:

> At first I was sure that I'd never get Tailwind Intellisense in Monaco, given how difficult it is to configure anything in it. Then I discovered this repo and it felt like magic.

If you’re up for a challenge, I more than welcome a PR to update to the latest version of Tailwind.

---

I suggest you have a look at https://github.com/remcohaszing/monaco-tailwindcss/blob/v0.6.1/examples/demo/src/tailwindcssplugin.worker.js. `prepareTailwindConfig()` accepts the value passed in https://github.com/remcohaszing/monaco-tailwindcss/blob/v0.6.1/examples/demo/src/index.ts#L41 and https://github.com/remcohaszing/monaco-tailwindcss/blob/v0.6.1/examples/demo/src/index.ts#L267. As a result, you can see the color of `bg-lava` on https://monaco-tailwindcss.js.org/#html, change it in the Tailwind configuration to say `#00ff00`, and see the change is reflected if you navigate back to the HTML tab.

You can use the same method to define your own config resolution. The API is deliberately loose so you can define your own things. You could to something like this for example:

```ts
const tailwindConfig = {
  // ...
  plugins: [],

  // Just define your custom things here.
  api: {
    addUtilities: {
      '.text-accent': {color: 'var(--brand-12)'},
      '.text-muted': {color: 'var(--gray-11)'},      
    }
    // addBase, addDefaults, addComponents, ...
  }
}

configureMonacoTailwindcss(monaco, {tailwindConfig})
```

I’ll leave the type casting exercise up to you.

**Comment by wkd-kapsule** [comment]
> If you’re up for a challenge, I more than welcome a PR to update to the latest version of Tailwind.

I'll give it a shot!

Concerning my problem, I already know how to adjust a config for example to add an `ocean-*` color palette. This isn't an issue. What I want to achieve is adding custom classes that don't belong to an existing `theme` property. I can easily modify the `color` or `fontSize` property (I think Tailwind calls them `group`) but what if I need to add a new group, like `button`? Where each variant would hold several css properties. With Tailwind, I have to use `addUtilities` so that the classes are pick up by the Intellisense extension. Here, I need an other way to do that as I can't use `addUtilities` within the config I pass to `configureMonacoTailwindcss`.

I'm probably wrong but from what you said and the examples you linked, I could just do something like this:

``` typescript
const tailwindConfig = {
  // ...
  plugins: [],

  // this
  api: {
    addUtilities: {
      '.button': {
        color: '#someColor',
        backgroundColor: '#someColor',
        padding: '0.5rem 1rem',
        fontSize: '1rem'
      },
      '.button-sm': {
        color: '#someColor',
        backgroundColor: '#someColor',
        padding: '0.25rem 0.5rem',
        fontSize: '0.8rem'
      },      
    }

  // or maybe this without nesting inside api.addUtilities?
  '.button': {
    color: '#someColor',
    backgroundColor: '#someColor',
    padding: '0.5rem 1rem',
    fontSize: '1rem'
  },
  '.button-sm': {
    color: '#someColor',
    backgroundColor: '#someColor',
    padding: '0.25rem 0.5rem',
    fontSize: '0.8rem'
  },      
}

configureMonacoTailwindcss(monaco, {tailwindConfig})
```

**Comment by wkd-kapsule** [comment]
Alright, I think it's done! It should be fixed by #93
